File: ios/Gem/Services/DeviceSyncPreflight.swift
// Copyright (c). Gem Wallet. All rights reserved.

import GemstoneServices
import Gemstone

final class DeviceSyncPreflight: GemWalletRequestPreflight, @unchecked Sendable {
    private let deviceService: any GemDeviceServiceProtocol

    init(deviceService: any GemDeviceServiceProtocol) {
        self.deviceService = deviceService
    }

    func prepare() async throws {
        try await deviceService.synchronizeIfNeeded()
    }
}

File: android/app/src/main/kotlin/com/gemwallet/android/services/DeviceSyncPreflight.kt
package com.gemwallet.android.services

import dagger.Lazy
import uniffi.gemstone.GemWalletRequestPreflight
import uniffi.gemstone.GemDeviceService

class DeviceSyncPreflight(
    private val deviceService: Lazy<GemDeviceService>,
) : GemWalletRequestPreflight {
    override suspend fun prepare() = deviceService.get().synchronizeIfNeeded()
}
